java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.database
+
+import android.content.ContentValues
+import android.content.Context
+import android.database.sqlite.SQLiteDatabase
+import android.database.sqlite.SQLiteOpenHelper
+import androidx.annotation.VisibleForTesting
+import arcs.android.storage.transaction
+import arcs.android.storage.useTransaction
+import arcs.core.data.Entity
+import arcs.core.data.FieldType
+import arcs.core.data.PrimitiveType
+import arcs.core.data.Schema
+import arcs.core.storage.database.Database
+import arcs.core.storage.StorageKey
+import java.lang.IllegalArgumentException
+
+/** The Type ID that gets stored in the database. */
+typealias TypeId = Long
+
+class DatabaseImpl(
+    context: Context
+) : Database, SQLiteOpenHelper(
+    context,
+    DB_NAME,
+    /* cursorFactory = */ null,
+    DB_VERSION
+) {
+    /** Maps from schema hash to type ID (local copy of the 'types' table). */
+    @VisibleForTesting
+    val schemaTypeMap = lazy { loadTypes() }
```

Comment:
I think we're going to need to use a mutex to work with the schemaTypeMap.  (this DatabaseImpl should be thread-safe)

Reply:
Added TODO.